File: theSrc/scripts/Box.js
/* Box class */

import _ from 'lodash'
import * as d3 from 'd3'

class Box {
  static initClass () {
    this.widgetIndex = 0
    this.widgetName = 'Box'
  }

  constructor (el, stateChangedCallback) {
    this.id = `${this.widgetName}-${this.widgetIndex++}`
    this.rootElement = _.has(el, 'length') ? el[0] : el
    this.state = {}
    this.stateChangedCallback = stateChangedCallback
  }

  resize () {

  }

  setUserState (userState = {}) {
    return this
  }

  setConfig (config) {
    this.config = config
    return this
  }

  setWidth (width) {
    this.width = this.config.width ? this.config.width : width
    return this
  }

  setHeight (height) {
    this.height = this.config.height ? this.config.height : height
    return this
  }

  draw () {
    this._clearRootElement()
    let containerEl = d3.select(this.rootElement)
      .append('div')
      .attr('class', 'MetroContainer')

    let containerBG
    if (this.config.background_shape == 'Ellipse') {
      containerBG = containerEl
      .append('svg')
      .attr('class', 'MetroBoxBg')
      .attr('width', this.width + 'px')
      .attr('height', this.height + 'px')
    }
    
    let containerMain = containerEl
      .append('div')
      .attr('class', 'MetroBox')
      .style('position', 'Absolute')
      .style('top', 0)
      .style('width', this.width + 'px')
      .style('height', this.height + 'px')

    let box = containerMain.append('div')
      .attr('class', 'MetroBoxInner')
      .style('display', 'table')
      .style('width', this.config.border_width == 0 ? this.width + 'px' : (this.width-this.config.border_width*2) + 'px')
      .style('height', this.config.border_width == 0 ? this.height + 'px' : (this.height-this.config.border_width*2) + 'px')
      .style('text-align', this.config.horizontal_align)
      .style('white-space', () => {
        if (this.config.wrap_text) {
          return 'normal'
        } else {
          return 'nowrap'
        }
      })

    if (this.config.background_shape == 'Rectangle') {
      box.style('background-color', this.config.background_color)
      .style('border-style', this.config.border_style)
      .style('border-width', this.config.border_width + 'px')
      .style('border-color', this.config.border_color)

    } else if (this.config.background_shape == 'Ellipse'){
      containerBG.append('ellipse')
        .attr('class', 'BoxEllipse')
        .attr('cx',this.width/2 + 'px')
        .attr('cy',this.height/2 + 'px')
        .attr('rx',this.config.border_width == 0 ? this.width/2 + 'px' : (this.width-this.config.border_width*2)/2 + 'px')
        .attr('ry',this.config.border_width == 0 ? this.height/2 + 'px' : (this.height-this.config.border_width*2)/2 + 'px')
        .style('stroke', this.config.border_color)
        .style('stroke-width', this.config.border_width + 'px')
        .style('stroke-dasharray', ()=>{
          if (this.config.border_style == 'Solid') {
            return undefined
          } else if (this.config.border_style == 'Dashed') {
            return '7,7'
          } else {
            console.log('Error: Elliptic background only support Solid or Dashed borders')
            return undefined
          }
        })
        .style('fill', this.config.background_color)

    } else {
      console.log('Incorrect background color input')
    }


    if (this.config.text) {
      let box1 = box.append('tspan')
        .attr('x', 0)
        .attr('y', 0)
        .attr('dy', 0)
        .style('display', 'table-cell')
        .style('vertical-align', this.config.vertical_align)
        .style('font-family', this.config.font_family + ', sans-serif')
        .style('font-size', this.config.font_size + 'pt')
        .style('color', this.config.font_color)
        .style('font-weight', this.config.font_bold ? 'bold' : 'normal')
        .style('font-style', this.config.font_italic ? 'italic' : 'normal')
        .style('text-decoration', () => {
          let dec = 'none'
          if (this.config.font_strikethrough) {
            return 'line-through'
          }
            
          if (this.config.font_underline) {
            return 'underline'
          }
        })

      if (this.config.as_html) {
        box1.html(this.config.text)
      } else {
        box1.text(this.config.text)
      }
    }
  }

  _clearRootElement () {
    d3.select(this.rootElement)
      .selectAll('div')
      .remove()
  }
}

module.exports = Box

File: R/Box.R
#' Box
#'
#' Creates an html Div container box with multiple features.
#' 
#' @param text character. Adding text to the box.
#' @param horizontal_align horizontal alignment of \code{text}. Possible values are c("left","center","right"). Defaults to \code{"left"}.
#' @param vertical_align vertical alignment of \code{text}. Possible values are \code{c("top","middle","bottom")}. Defaults to \code{"top"}.
#' @param wrap_text logical. Whether text wrapping is enabled. Defaults to \code{TRUE}.
#' @param background_color HTML color string. Sets the background color of the box. Defaults to \code{"Transparent"}.
#' @param font_family HTML font family string. Sets the font family of \code{text}. Defaults to \code{"sans-serif"}.
#' @param font_size integer. Sets the font size of \code{text} in pixels. Defaults to 11.
#' @param font_color HTML color string. Sets the font color of \code{text}. Defaults to \code{"black"}.
#' @param font_bold logical. Whether to bold \code{text}. Defaults to \code{FALSE}.
#' @param font_italic logical. Whether to have italic \code{text}. Defaults to \code{FALSE}.
#' @param font_underline logical. Whether to underline \code{text}. Defaults to \code{FALSE}.
#' @param font_strikethrough logical. Whether to draw a horizontal line across \code{text}. Defaults to code{FALSE}.
#' @param border_width integer. Sets the border width in pixels. Defaults to 0.
#' @param border_color HTML color string. Sets the border color. Defaults to "Transparent".
#' @param border_style HTML border style string. Sets border style. Defaults to "Solid".
#' @param as_html logical. Whether to interpret \code{text} as raw html. Defaults to \code{FALSE}.
#' @param width integer. Width of the widget in pixels. Defaults to NULL, which automatically calculates the width based on window size.
#' @param height integer. Height of the widget in pixels. Defaults to NULL, which automatically calculates the height based on window size.

#' @import htmlwidgets
#' @export

Box <- function(
    text = NULL,
    horizontal.align = "left",
    vertical.align = "top",
    wrap.text = TRUE,
    background.color = "Transparent",
    font.family = "sans-serif",
    font.size = 11,
    font.color = "black",
    font.bold = FALSE,
    font.italic = FALSE,
    font.underline = FALSE,
    font.strikethrough = FALSE,
    border.width = 0,
    border.color = "Transparent",
    border.style = "Solid",
    as.html = FALSE,
    width = NULL,
    height = NULL
) {
    
    payload <- list(
        text = text,
        horizontal_align = horizontal.align,
        vertical_align = vertical.align,
        wrap_text = wrap.text,
        background_color = background.color,
        font_family = font.family,
        font_size = font.size,
        font_color = font.color,
        font_bold = font.bold,
        font_italic = font.italic,
        font_underline = font.underline,
        font_strikethrough = font.strikethrough,
        border_width = border.width,
        border_color = border.color,
        border_style = border.style,
        as_html = as.html,
        class = "box",
        width = width,
        height = height
    )

    htmlwidgets::createWidget(
        name = 'rhtmlMetro',
        payload,
        width = width,
        height = height,
        sizingPolicy = htmlwidgets::sizingPolicy(
            browser.padding = 5,
            viewer.padding = 5,
            browser.fill = TRUE,
            viewer.fill = TRUE
        ),
        package = 'rhtmlMetro'
    )
}



